File: 백준/Platinum/6549. 히스토그램에서 가장 큰 직사각형/히스토그램에서 가장 큰 직사각형.py
def solution(hi: list[int]):
    hi.append(0)
    stack = []
    maxArea = 0

    for idx in range(len(hi)):
        currentIdx = idx
        
        while stack and stack[-1][1] >= hi[idx]:
            currentIdx, currentHeight = stack.pop()
            maxArea = max(maxArea, (idx-currentIdx) * currentHeight)
        
        stack.append ( (currentIdx, hi[idx]) )
        
    print(maxArea)
        

while True:
    line = list(map(int, input().split()))

    n = line[0]

    if n == 0:
        break
        
    squares = line[1:]

    solution(squares)

File: 백준/Platinum/6549. 히스토그램에서 가장 큰 직사각형/히스토그램에서 가장 큰 직사각형.swift
func solution(_ squares: inout [Int]) {
    squares.append(0)
    
    var maxArea: Int = 0
    
    var stack = [(Int, Int)]()
    for idx in 1..<squares.count {
        var currentIdx: Int = idx
        let currentHeight: Int = squares[idx]

        while let (topIdx, topHeight): (Int, Int) = stack.last {
            if topHeight >= currentHeight {
                currentIdx = topIdx
                stack.removeLast()

                maxArea = max(maxArea, (idx-currentIdx) * topHeight)
            } else {
                break
            }
        }
        
        stack.append( (currentIdx, currentHeight) )
    }

    print(maxArea)
}

while let input: String = readLine() {
    var squares: [Int] = input.split(separator: " ").compactMap { Int($0) }

    let n: Int = squares[0]
    if n == 0 {
        break
    }
    
    solution(&squares)
}